api/client/swarm/init.go

```diff
@@ -45,19 +54,61 @@ func newInitCommand(dockerCli *client.DockerCli) *cobra.Command {
 func runInit(dockerCli *client.DockerCli, flags *pflag.FlagSet, opts initOptions) error {
 	client := dockerCli.Client()
 	ctx := context.Background()
+	spec := opts.swarmOptions.ToSpec()
+
+	var secret string
+	// If no secret was specified, we create a random one
+	if !flags.Changed("secret") {
+		var secretBytes [generatedSecretEntropyBytes]byte
+
+		if _, err := cryptorand.Read(secretBytes[:]); err != nil {
+			panic(fmt.Errorf("failed to read random bytes: %v", err))
+		}
+
+		var nn big.Int
+		nn.SetBytes(secretBytes[:])
+		secret = fmt.Sprintf("%0[1]*s", maxGeneratedSecretLength, nn.Text(generatedSecretBase))
+
+		fmt.Fprintf(dockerCli.Out(), "No node joining secret provided. Generated random secret:\n\t%s\n\n", secret)
+
+		for i := range spec.AcceptancePolicy.Policies {
+			spec.AcceptancePolicy.Policies[i].Secret = &secret
+		}
+	} else {
+		secret = opts.secret
+	}
```

Comment:
The secret is set in both blocks. In the first block it's

`secret = fmt.Sprintf("%0[1]*s", maxGeneratedSecretLength, nn.Text(generatedSecretBase))`

and in the second block it's

`secret = opts.secret`

In other words, if the user specifies a secret, we set the `secret` variable to that value. Otherwise, we generate one and set it to the generated value.

I guess you're right that the `if` case could set `opts.secret` directly. That didn't feel quite right to me, because the generated secret is not an option provided by the user.